pkg/datapath/iptables/iptables.go

```diff
@@ -1035,12 +1046,56 @@ func (m *IptablesManager) installForwardChainRulesIpX(prog iptablesInterface, if
 	return nil
 }
 
+// AddToNodeIpset adds an IP address to the ipset for cluster nodes. It creates
+// the ipset if it doesn't already exist and doesn't error if either the ipset
+// or the IP already exist.
+func AddToNodeIpset(nodeIP net.IP) {
+	ciliumNodeIpset := ciliumNodeIpsetV4
+	if ip.IsIPv6(nodeIP) {
+		ciliumNodeIpset = ciliumNodeIpsetV6
+	}
+	log.WithField(logfields.IPAddr, nodeIP.String()).Debugf("Adding IP to %s ipset", ciliumNodeIpset)
+	createIpset(ciliumNodeIpset, ip.IsIPv6(nodeIP))
+	progArgs := []string{"add", ciliumNodeIpset, nodeIP.String(), "-exist"}
+	ipset.runProg(progArgs, false)
+}
+
+// RemoveFromBodeIpset removes an IP address from the ipset for cluster nodes.
+func RemoveFromNodeIpset(nodeIP net.IP) {
+	ciliumNodeIpset := ciliumNodeIpsetV4
+	if ip.IsIPv6(nodeIP) {
+		ciliumNodeIpset = ciliumNodeIpsetV6
+	}
+	log.WithField(logfields.IPAddr, nodeIP.String()).Debugf("Removing IP to %s ipset", ciliumNodeIpset)
+	progArgs := []string{"del", ciliumNodeIpset, nodeIP.String()}
+	ipset.runProg(progArgs, false)
+}
+
 func (m *IptablesManager) installMasqueradeRules(prog iptablesInterface, ifName, localDeliveryInterface,
 	snatDstExclusionCIDR, allocRange, hostMasqueradeIP string) error {
+	if option.Config.Tunnel == option.TunnelDisabled {
+		// Exclude traffic to nodes from masquerade.
+		if err := createIpset(prog.getIpset(), prog.getProg() == "ip6tables"); err != nil {
+			return err
+		}
+		progArgs := append(
+			m.waitArgs,
+			"-t", "nat",
+			"-A", ciliumPostNatChain,
```

Comment:
Just thinking out loud: should we append or insert here? I would assume here the nat rules are clear when we enter and the masquerading rules are added after this?

Reply:
> I would assume here the nat rules are clear when we enter and the masquerading rules are added after this?

Yes, I think that's the assumption. Other rules are appended as well. IIRC, we have code to first clear all old rules (see `removeOldRules`).

Reply:
Cool thanks for answering.